apollo-router/src/plugins/telemetry/config_new/events.rs

```diff
@@ -244,49 +195,39 @@ where
     Attributes: Selectors<Request, Response, EventResponse> + Default,
     Sel: Selector<Request = Request, Response = Response> + Debug,
 {
-    pub(super) request: StandardEvent<Sel>,
-    pub(super) response: StandardEvent<Sel>,
-    pub(super) error: StandardEvent<Sel>,
+    pub(super) request: Option<StandardEvent<Sel>>,
+    pub(super) response: Option<StandardEvent<Sel>>,
+    pub(super) error: Option<StandardEvent<Sel>>,
     pub(super) custom: Vec<CustomEvent<Request, Response, EventResponse, Attributes, Sel>>,
 }
 
-impl Instrumented
-    for CustomEvents<router::Request, router::Response, (), RouterAttributes, RouterSelector>
-{
-    type Request = router::Request;
-    type Response = router::Response;
-    type EventResponse = ();
-
-    fn on_request(&self, request: &Self::Request) {
-        if self.request.level() != EventLevel::Off {
-            if let Some(condition) = self.request.condition() {
-                if condition.lock().evaluate_request(request) != Some(true) {
-                    return;
-                }
+impl CustomEvents<router::Request, router::Response, (), RouterAttributes, RouterSelector> {
+    pub(crate) fn on_request(&mut self, request: &router::Request) {
+        if let Some(request_event) = &mut self.request {
+            if request_event.condition.evaluate_request(request) != Some(true) {
+                return;
             }
 
             request
                 .context
                 .extensions()
-                .with_lock(|ext| ext.insert(DisplayRouterRequest(self.request.level())));
+                .with_lock(|ext| ext.insert(DisplayRouterRequest(request_event.level)));
         }
-        if self.response.level() != EventLevel::Off {
+        if self.response.is_some() {
             request
                 .context
                 .extensions()
                 .with_lock(|ext| ext.insert(DisplayRouterResponse(true)));
         }
-        for custom_event in &self.custom {
+        for custom_event in &mut self.custom {
             custom_event.on_request(request);
         }
     }
 
-    fn on_response(&self, response: &Self::Response) {
-        if self.response.level() != EventLevel::Off {
-            if let Some(condition) = self.response.condition() {
-                if !condition.lock().evaluate_response(response) {
-                    return;
-                }
+    pub(crate) fn on_response(&mut self, response: &router::Response) {
+        if let Some(response_event) = &mut self.response {
+            if !response_event.condition.evaluate_response(response) {
+                return;
```

Comment:
note: this PR doesn't fix early exit issues I talked about in PR description.
I will fix them in separate PR